pennylane/transforms/control.py

```diff
@@ -0,0 +1,191 @@
+    Returns:
+        A new QuantumTape with the controlled operations.
+    """
+    with QuantumTape(do_queue=False) as new_tape:
+        for op in tape.operations:
+            if op.__class__ in CONTROL_MAPS:
+                # Create the controlled version of the operation
+                # and add that the to the tape context.
+                # NOTE(chase): Should we use type(op) here instead?
+                CONTROL_MAPS[op.__class__](op, control_wire)
+            else:
+                raise NotImplementedError(
+                    f"Control transform of operation {type(op)} is not registered."
+                    "You can define the control transform by using qml.register_control"
+                    "See documentation for more details."
+                )
+    return new_tape
+
+
+class ControlledOperation(Operation):
+    """A Controlled Operation.
+
+    Unless you are a Pennylane plugin developer, **you should NOT directly use this class**,
+    instead, use the ``qml.ctrl`` method.
+
+    The ``ControlledOperation`` is a container class that defines a set of operations that
+    should by applied relative to a single control wire or a list of control wires.
+
+    Certain simulators and quantum computers can take advantage of the controlled gate sparsity,
+    while other devices must rely on the op-by-op decomposition defined by the ``op.expand``
+    method.
+
+    Attributes:
+        subtape (QuantumTape): The tape that defines the underlying operation.
+        control_wires (Wires): The control wires.
+
+    Args:
+        tape: A QuantumTape. This tape defines the unitary that should be applied relative
+            to the control wires.
+        control_wires: A wire or set of wires.
+    """
+
+    par_domain = "A"
+    num_wires = AnyWires
+    num_params = property(lambda self: self.subtape.num_params)
+
+    def __init__(self, tape, control_wires, do_queue=True):
+        self.subtape = tape
+        self.control_wires = Wires(control_wires)
+        wires = self.control_wires + tape.wires
+        super().__init__(*tape.get_parameters(), wires=wires, do_queue=do_queue)
+
+    def expand(self):
+        tape = self.subtape
+        for wire in self.control_wires:
+            tape = expand_with_control(tape, wire)
+        # TODO(chase): Do we need to re-queue these ops?
+        return tape
+
+    def adjoint(self, do_queue=False):
+        def requeue_tape(tape):
+            for op in tape.operations:
+                op.queue()
+
+        with QuantumTape(do_queue=False) as new_tape:
+            # Execute all ops adjointed.
+            adjoint(requeue_tape)(self.subtape)
+        return ControlledOperation(new_tape, self.control_wires, do_queue=do_queue)
+
+
+register_control(
+    ControlledOperation,
+    lambda op, wires: ControlledOperation(
+        tape=op.subtape,
+        control_wires=Wires(wires) + op.control_wires,
+    ),
+)
+
+
+def ctrl(fn, control):

```

Comment:
I am a bit confused about the `fn`. For example, does that mean I cannot do this?

```
def circuit():
   qml.ctrl(qml.RX)
   return ...
```

That would be the first thing I'd try as a user :)

Reply:
Nope, it would work the same way that `qml.grad(fn)(*args)`, `qml.draw(fn)(*args)` etc. work:

```python
def circuit():
   qml.ctrl(qml.PauliX, 0)(wires=1)
   return ...
```

Definitely might be confusing for new users, but as we have more functional transforms, hopefully this pattern should become familiar across PL

Reply:
I also want to avoid object modification. I don't see how we could ever get `qml.ctrl(qml.PauliX(wires=1))` working without either inplace modifying the tape or the PauliX operation.